BlackBox of `imap_go` is broken

The following:
```haskell
module TXYZ where

import Clash.Prelude

topEntity :: Vec 3 (Index 3)
topEntity = case indicesI @3 of
  (Cons x xs) -> Cons maxBound xs
```
produces:
```
TXYZ.hs:6:1: error:
    
    Clash.Netlist.BlackBox(212): Couldn't instantiate blackbox for Clash.Sized.Vector.imap_go. Verification procedure reported:
    
    Blackbox requested instantiation of function at argument 1, but BlackBoxContext did not contain one.
    
    The source location of the error is not exact, only indicative, as it is acquired 
    after optimizations. The actual location of the error can be in a function that is 
    inlined. To prevent inlining of those functions, annotate them with a NOINLINE pragma.
  |
6 | topEntity = case indicesI @3 of
  | ^^^^^^^^^
```
This is due to https://github.com/clash-lang/clash-compiler/commit/a1dacb46446b3c3c841dc4b174a3b1a719632717 where I swiched the arguments in the compile-time evaluation rule for `imap_go`, but not for the blackboxes.

--- a/changelog/2025-04-28T14_17_05+02_00_fix2542
+++ b/changelog/2025-04-28T14_17_05+02_00_fix2542
@@ -1,0 +1,1 @@
+FIXED: Fixed laziness issue in internal black box `imap_go` [#2542](https://github.com/clash-lang/clash-compiler/issues/2542)

--- a/clash-lib/prims/systemverilog/Clash_Sized_Vector.primitives.yaml
+++ b/clash-lib/prims/systemverilog/Clash_Sized_Vector.primitives.yaml
@@ -1,29 +1,25 @@
 - BlackBox:
     name: Clash.Sized.Vector.head
     kind: Expression
-    type: 'head :: Vec (n
-      + 1) a -> a'
+    type: 'head :: Vec (n + 1) a -> a'
     template: ~FROMBV[~VAR[vec][0][\0\]][~TYPO]
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.tail
     kind: Expression
-    type: 'tail :: Vec (n
-      + 1) a -> Vec n a'
+    type: 'tail :: Vec (n + 1) a -> Vec n a'
     template: '~VAR[vec][0][1 : $high(~VAR[vec][0])]'
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.last
     kind: Expression
-    type: Vec (n + 1) a ->
-      a
+    type: Vec (n + 1) a -> a
     template: ~FROMBV[~VAR[vec][0][\$high(~VAR[vec][0])\]][~TYPO]
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.init
     kind: Expression
-    type: Vec (n + 1) a ->
-      Vec n a
+    type: Vec (n + 1) a -> Vec n a
     template: '~VAR[vec][0][0 : $high(~VAR[vec][0]) - 1]'
     workInfo: Never
 - BlackBox:
@@ -49,15 +45,13 @@
 - BlackBox:
     name: Clash.Sized.Vector.++
     kind: Expression
-    type: '(++) :: Vec n a
-      -> Vec m a -> Vec (n + m) a'
+    type: '(++) :: Vec n a -> Vec m a -> Vec (n + m) a'
     template: ~FROMBV[{~TOBV[~ARG[0]][~TYP[0]],~TOBV[~ARG[1]][~TYP[1]]}][~TYPO]
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.concat
     kind: Declaration
-    type: 'concat :: Vec n
-      (Vec m a) -> Vec (n * m) a'
+    type: 'concat :: Vec n (Vec m a) -> Vec (n * m) a'
     template: |-
       // concat begin
       genvar ~GENSYM[n][1];
@@ -71,8 +65,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.splitAt
     kind: Declaration
-    type: 'splitAt :: SNat
-      m -> Vec (m + n) a -> (Vec m a, Vec n a)'
+    type: 'splitAt :: SNat m -> Vec (m + n) a -> (Vec m a, Vec n a)'
     template: |-
       // splitAt begin~IF~LENGTH[~TYPO]~THEN
       assign ~RESULT = ~ARG[1];~ELSE
@@ -110,8 +103,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.map
     kind: Declaration
-    type: 'map :: (a -> b)
-      -> Vec n a -> Vec n b'
+    type: 'map :: (a -> b) -> Vec n a -> Vec n b'
     template: |-
       // map begin
       genvar ~GENSYM[n][1];
@@ -132,8 +124,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.imap
     kind: Declaration
-    type: 'imap :: KnownNat
-      n => (Index n -> a -> b) -> Vec n a -> Vec n b'
+    type: 'imap :: KnownNat n => (Index n -> a -> b) -> Vec n a -> Vec n b'
     template: |-
       // imap begin
       genvar ~GENSYM[n][1];
@@ -157,8 +148,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.imap_go
     kind: Declaration
-    type: 'imap_go :: Index
-      n -> (Index n -> a -> b) -> Vec m a -> Vec m b'
+    type: 'imap_go :: Index n -> (Index n -> a -> b) -> Vec m a -> Vec m b'
     template: |-
       // imap begin
       genvar ~GENSYM[n][1];
@@ -182,8 +172,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.zipWith
     kind: Declaration
-    type: 'zipWith :: (a ->
-      b -> c) -> Vec n a -> Vec n b -> Vec n c'
+    type: 'zipWith :: (a -> b -> c) -> Vec n a -> Vec n b -> Vec n c'
     template: |-
       // zipWith begin
       genvar ~GENSYM[n][2];
@@ -207,8 +196,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.foldr
     kind: Declaration
-    type: 'foldr :: (a ->
-      b -> b) -> b -> Vec n a -> b'
+    type: 'foldr :: (a -> b -> b) -> b -> Vec n a -> b'
     template: |-
       // foldr start~IF ~LENGTH[~TYP[2]] ~THEN
       ~SIGDO[~GENSYM[intermediate][0]] [0:~LENGTH[~TYP[2]]];
@@ -235,15 +223,13 @@
 - BlackBox:
     name: Clash.Sized.Vector.index_int
     kind: Expression
-    type: 'index_int :: KnownNat
-      n => Vec n a -> Int -> a'
+    type: 'index_int :: KnownNat n => Vec n a -> Int -> a'
     template: ~IF~SIZE[~TYP[1]]~THEN~FROMBV[~VAR[vec][1][\~ARG[2]\]][~TYPO]~ELSE~ERRORO~FI
 - BlackBox:
     name: Clash.Sized.Vector.replace_int
     kind: Declaration
     outputUsage: Blocking
-    type: 'replace_int ::
-      KnownNat n => Vec n a -> Int -> a -> Vec n a'
+    type: 'replace_int :: KnownNat n => Vec n a -> Int -> a -> Vec n a'
     template: |-
       // replaceVec start
       always_comb begin
@@ -254,15 +240,13 @@
 - BlackBox:
     name: Clash.Sized.Vector.replicate
     kind: Expression
-    type: 'replicate :: SNat
-      n -> a -> Vec n a'
-    template: '''{~LIT[0] {~TOBV[~ARG[1]][~TYP[1]]}}'
+    type: 'replicate :: SNat n -> a -> Vec n a'
+    template: "'{~LIT[0] {~TOBV[~ARG[1]][~TYP[1]]}}"
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.transpose
     kind: Declaration
-    type: 'transpose :: KnownNat
-      n => Vec m (Vec n a) -> Vec n (Vec m a)'
+    type: 'transpose :: KnownNat n => Vec m (Vec n a) -> Vec n (Vec m a)'
     template: |-
       // transpose begin
       genvar ~GENSYM[row_index][1];
@@ -280,8 +264,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.reverse
     kind: Declaration
-    type: 'reverse :: Vec
-      n a -> Vec n a'
+    type: 'reverse :: Vec n a -> Vec n a'
     template: |-
       // reverse begin
       genvar ~GENSYM[n][1];
@@ -313,8 +296,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.rotateLeftS
     kind: Declaration
-    type: 'rotateLeftS ::
-      KnownNat n => Vec n a -> SNat d -> Vec n a'
+    type: 'rotateLeftS :: KnownNat n => Vec n a -> SNat d -> Vec n a'
     template: |-
       // rotateLeftS begin
       localparam ~GENSYM[shift_amount][2] = ~LIT[2] % ~LIT[0];
@@ -332,8 +314,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.rotateRightS
     kind: Declaration
-    type: 'rotateRightS ::
-      KnownNat n => Vec n a -> SNat d -> Vec n a'
+    type: 'rotateRightS :: KnownNat n => Vec n a -> SNat d -> Vec n a'
     template: |-
       // rotateRightS begin
       localparam ~GENSYM[shift_amount][2] = ~LIT[2] % ~LIT[0];

--- a/clash-lib/prims/systemverilog/Clash_Sized_Vector.primitives.yaml
+++ b/clash-lib/prims/systemverilog/Clash_Sized_Vector.primitives.yaml
@@ -148,21 +148,21 @@
 - BlackBox:
     name: Clash.Sized.Vector.imap_go
     kind: Declaration
-    type: 'imap_go :: Index n -> (Index n -> a -> b) -> Vec m a -> Vec m b'
+    type: 'imap_go :: (Index n -> a -> b) -> Vec m a -> Index n -> Vec m b'
     template: |-
       // imap begin
       genvar ~GENSYM[n][1];
       ~GENERATE
       for (~SYM[1]=0; ~SYM[1] < $size(~RESULT); ~SYM[1] = ~SYM[1] + 1) begin : ~GENSYM[imap][2]
-        ~TYP[0] ~GENSYM[i][3];
-        assign ~SYM[3] = ~SYM[1] + ~ARG[0];~IF~SIZE[~TYP[2]]~THEN
-        ~TYPEL[~TYP[2]] ~GENSYM[imap_in][4];
-        assign ~SYM[4] = ~FROMBV[~VAR[vec][2][\~SYM[1]\]][~TYPEL[~TYP[2]]];~ELSE ~FI
+        ~TYP[2] ~GENSYM[i][3];
+        assign ~SYM[3] = ~SYM[1] + ~ARG[2];~IF~SIZE[~TYP[1]]~THEN
+        ~TYPEL[~TYP[1]] ~GENSYM[imap_in][4];
+        assign ~SYM[4] = ~FROMBV[~VAR[vec][1][\~SYM[1]\]][~TYPEL[~TYP[1]]];~ELSE ~FI
         ~TYPEL[~TYPO] ~GENSYM[imap_out][5];
-        ~INST 1
+        ~INST 0
           ~OUTPUT <= ~SYM[5]~ ~TYPEL[~TYPO]~
-          ~INPUT  <= ~SYM[3]~ ~TYP[0]~
-          ~INPUT  <= ~SYM[4]~ ~TYPEL[~TYP[2]]~
+          ~INPUT  <= ~SYM[3]~ ~TYP[2]~
+          ~INPUT  <= ~SYM[4]~ ~TYPEL[~TYP[1]]~
         ~INST
         assign ~RESULT[~SYM[1]] = ~TOBV[~SYM[5]][~TYPEL[~TYPO]];
       end

--- a/clash-lib/prims/verilog/Clash_Sized_Vector.primitives.yaml
+++ b/clash-lib/prims/verilog/Clash_Sized_Vector.primitives.yaml
@@ -1,29 +1,25 @@
 - BlackBox:
     name: Clash.Sized.Vector.head
     kind: Expression
-    type: 'head :: Vec (n
-      + 1) a -> a'
+    type: 'head :: Vec (n + 1) a -> a'
     template: '~FROMBV[~VAR[vec][0][\~SIZE[~TYP[0]]-1 -: ~SIZE[~TYPO]\]][~TYPO]'
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.tail
     kind: Expression
-    type: 'tail :: Vec (n
-      + 1) a -> Vec n a'
+    type: 'tail :: Vec (n + 1) a -> Vec n a'
     template: '~VAR[vec][0][~SIZE[~TYPO]-1 : 0]'
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.last
     kind: Expression
-    type: Vec (n + 1) a ->
-      a
+    type: Vec (n + 1) a -> a
     template: ~FROMBV[~VAR[vec][0][\~SIZE[~TYPO]-1:0\]][~TYPO]
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.init
     kind: Expression
-    type: Vec (n + 1) a ->
-      Vec n a
+    type: Vec (n + 1) a -> Vec n a
     template: '~VAR[vec][0][~SIZE[~TYP[0]]-1 : ~SIZE[~TYPEL[~TYP[0]]]]'
     workInfo: Never
 - BlackBox:
@@ -57,22 +53,19 @@
 - BlackBox:
     name: Clash.Sized.Vector.++
     kind: Expression
-    type: '(++) :: Vec n a
-      -> Vec m a -> Vec (n + m) a'
+    type: '(++) :: Vec n a -> Vec m a -> Vec (n + m) a'
     template: '{~ARG[0],~ARG[1]}'
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.concat
     kind: Expression
-    type: 'concat :: Vec n
-      (Vec m a) -> Vec (n * m) a'
+    type: 'concat :: Vec n (Vec m a) -> Vec (n * m) a'
     template: ~ARG[0]
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.splitAt
     kind: Expression
-    type: 'splitAt :: SNat
-      m -> Vec (m + n) a -> (Vec m a, Vec n a)'
+    type: 'splitAt :: SNat m -> Vec (m + n) a -> (Vec m a, Vec n a)'
     template: ~ARG[1]
     workInfo: Never
 - BlackBox:
@@ -88,8 +81,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.map
     kind: Declaration
-    type: 'map :: (a -> b)
-      -> Vec n a -> Vec n b'
+    type: 'map :: (a -> b) -> Vec n a -> Vec n b'
     template: |-
       // map begin
       genvar ~GENSYM[i][1];
@@ -110,8 +102,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.imap
     kind: Declaration
-    type: 'imap :: KnownNat
-      n => (Index n -> a -> b) -> Vec n a -> Vec n b'
+    type: 'imap :: KnownNat n => (Index n -> a -> b) -> Vec n a -> Vec n b'
     template: |-
       // imap begin
       genvar ~GENSYM[i][1];
@@ -136,8 +127,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.imap_go
     kind: Declaration
-    type: 'imap :: Index n
-      -> (Index n -> a -> b) -> Vec m a -> Vec m b'
+    type: 'imap :: Index n -> (Index n -> a -> b) -> Vec m a -> Vec m b'
     template: |-
       // imap begin
       genvar ~GENSYM[i][1];
@@ -162,8 +152,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.zipWith
     kind: Declaration
-    type: 'zipWith :: (a ->
-      b -> c) -> Vec n a -> Vec n b -> Vec n c'
+    type: 'zipWith :: (a -> b -> c) -> Vec n a -> Vec n b -> Vec n c'
     template: |-
       // zipWith start
       genvar ~GENSYM[i][2];
@@ -187,8 +176,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.foldr
     kind: Declaration
-    type: 'foldr :: (a ->
-      b -> b) -> b -> Vec n a -> b'
+    type: 'foldr :: (a -> b -> b) -> b -> Vec n a -> b'
     template: |-
       // foldr start~IF ~LENGTH[~TYP[2]] ~THEN
       wire ~TYPO ~GENSYM[intermediate][0] [0:~LENGTH[~TYP[2]]];
@@ -223,8 +211,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.replace_int
     kind: Declaration
-    type: 'replace_int ::
-      KnownNat n => Vec n a -> Int -> a -> Vec n a'
+    type: 'replace_int :: KnownNat n => Vec n a -> Int -> a -> Vec n a'
     template: |-
       // vector replace begin
       genvar ~GENSYM[i][0];
@@ -237,29 +224,25 @@
 - BlackBox:
     name: Clash.Sized.Vector.maxIndex
     kind: Expression
-    type: 'maxIndex :: KnownNat
-      n => Vec n a -> Int'
+    type: 'maxIndex :: KnownNat n => Vec n a -> Int'
     template: ~SIZE[~TYPO]'sd~LIT[0] - ~SIZE[~TYPO]'d1
     workInfo: Constant
 - BlackBox:
     name: Clash.Sized.Vector.length
     kind: Expression
-    type: 'length :: KnownNat
-      n => Vec n a -> Int'
+    type: 'length :: KnownNat n => Vec n a -> Int'
     template: ~SIZE[~TYPO]'sd~LIT[0]
     workInfo: Constant
 - BlackBox:
     name: Clash.Sized.Vector.replicate
     kind: Expression
-    type: 'replicate :: SNat
-      n -> a -> Vec n a'
+    type: 'replicate :: SNat n -> a -> Vec n a'
     template: '{~LIT[0] {~ARG[1]}}'
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.transpose
     kind: Declaration
-    type: 'transpose :: KnownNat
-      n => Vec m (Vec n a) -> Vec n (Vec m a)'
+    type: 'transpose :: KnownNat n => Vec m (Vec n a) -> Vec n (Vec m a)'
     template: |-
       // transpose begin
       genvar ~GENSYM[row_index][1];
@@ -276,8 +259,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.reverse
     kind: Declaration
-    type: 'reverse :: Vec
-      n a -> Vec n a'
+    type: 'reverse :: Vec n a -> Vec n a'
     template: |-
       // reverse begin
       genvar ~GENSYM[i][1];
@@ -309,8 +291,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.rotateLeftS
     kind: Declaration
-    type: 'rotateLeftS ::
-      KnownNat n => Vec n a -> SNat d -> Vec n a'
+    type: 'rotateLeftS :: KnownNat n => Vec n a -> SNat d -> Vec n a'
     template: |-
       // rotateLeftS begin
       localparam ~GENSYM[shift_amount][2] = ~LIT[2] % ~LIT[0];
@@ -329,8 +310,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.rotateRightS
     kind: Declaration
-    type: 'rotateRightS ::
-      KnownNat n => Vec n a -> SNat d -> Vec n a'
+    type: 'rotateRightS :: KnownNat n => Vec n a -> SNat d -> Vec n a'
     template: |-
       // rotateRightS begin
       localparam ~GENSYM[shift_amount][2] = ~LIT[2] % ~LIT[0];

--- a/clash-lib/prims/verilog/Clash_Sized_Vector.primitives.yaml
+++ b/clash-lib/prims/verilog/Clash_Sized_Vector.primitives.yaml
@@ -127,22 +127,22 @@
 - BlackBox:
     name: Clash.Sized.Vector.imap_go
     kind: Declaration
-    type: 'imap :: Index n -> (Index n -> a -> b) -> Vec m a -> Vec m b'
+    type: 'imap_go :: (Index n -> a -> b) -> Vec m a -> Index n -> Vec m b'
     template: |-
       // imap begin
       genvar ~GENSYM[i][1];
       ~GENERATE
       for (~SYM[1]=0; ~SYM[1] < ~LENGTH[~TYPO]; ~SYM[1] = ~SYM[1] + 1) begin : ~GENSYM[imap][2]
-        wire ~TYP[0] ~GENSYM[map_index][3];~IF~SIZE[~TYP[2]]~THEN
-        wire ~TYPEL[~TYP[2]] ~GENSYM[map_in][4];
-        assign ~SYM[4] = ~VAR[vec][2][~SYM[1]*~SIZE[~TYPEL[~TYP[2]]]+:~SIZE[~TYPEL[~TYP[2]]]];~ELSE ~FI
-        ~OUTPUTUSAGE[1] ~TYPEL[~TYPO] ~GENSYM[map_out][5];
+        wire ~TYP[2] ~GENSYM[map_index][3];~IF~SIZE[~TYP[1]]~THEN
+        wire ~TYPEL[~TYP[1]] ~GENSYM[map_in][4];
+        assign ~SYM[4] = ~VAR[vec][1][~SYM[1]*~SIZE[~TYPEL[~TYP[1]]]+:~SIZE[~TYPEL[~TYP[1]]]];~ELSE ~FI
+        ~OUTPUTUSAGE[0] ~TYPEL[~TYPO] ~GENSYM[map_out][5];
 
-        assign ~SYM[3] = ~SIZE[~TYP[0]]'d~MAXINDEX[~TYPO] - ~SYM[1][0+:~SIZE[~TYP[0]]] + ~ARG[0];
-        ~INST 1
+        assign ~SYM[3] = ~SIZE[~TYP[2]]'d~MAXINDEX[~TYPO] - ~SYM[1][0+:~SIZE[~TYP[2]]] + ~ARG[2];
+        ~INST 0
           ~OUTPUT <= ~SYM[5]~ ~TYPEL[~TYPO]~
-          ~INPUT  <= ~SYM[3]~ ~TYP[0]~
-          ~INPUT  <= ~SYM[4]~ ~TYPEL[~TYP[2]]~
+          ~INPUT  <= ~SYM[3]~ ~TYP[2]~
+          ~INPUT  <= ~SYM[4]~ ~TYPEL[~TYP[1]]~
         ~INST
         assign ~RESULT[~SYM[1]*~SIZE[~TYPEL[~TYPO]]+:~SIZE[~TYPEL[~TYPO]]] = ~SYM[5];
       end

--- a/clash-lib/prims/vhdl/Clash_Sized_Vector.primitives.yaml
+++ b/clash-lib/prims/vhdl/Clash_Sized_Vector.primitives.yaml
@@ -1,31 +1,27 @@
 - BlackBox:
     name: Clash.Sized.Vector.head
     kind: Expression
-    type: 'head :: Vec (n
-      + 1) a -> a'
+    type: 'head :: Vec (n + 1) a -> a'
     template: ~IF ~VIVADO ~THEN ~TYPMO'(fromSLV(~VAR[vec][0](0))) ~ELSE ~VAR[vec][0](0)
       ~FI
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.tail
     kind: Expression
-    type: 'tail :: Vec (n
-      + 1) a -> Vec n a'
+    type: 'tail :: Vec (n + 1) a -> Vec n a'
     template: ~VAR[vec][0](1 to ~VAR[vec][0]'high)
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.last
     kind: Expression
-    type: Vec (n + 1) a ->
-      a
+    type: Vec (n + 1) a -> a
     template: ~IF ~VIVADO ~THEN ~TYPMO'(fromSLV(~VAR[vec][0](~VAR[vec][0]'high)))
       ~ELSE ~VAR[vec][0](~VAR[vec][0]'high) ~FI
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.init
     kind: Expression
-    type: Vec (n + 1) a ->
-      Vec n a
+    type: Vec (n + 1) a -> Vec n a
     template: ~VAR[vec][0](0 to ~VAR[vec][0]'high - 1)
     workInfo: Never
 - BlackBox:
@@ -48,15 +44,13 @@
 - BlackBox:
     name: Clash.Sized.Vector.++
     kind: Expression
-    type: '(++) :: Vec n a
-      -> Vec m a -> Vec (n + m) a'
+    type: '(++) :: Vec n a -> Vec m a -> Vec (n + m) a'
     template: ~TYPMO'(~TYPM[0]'(~ARG[0]) & ~TYPM[1]'(~ARG[1]))
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.concat
     kind: Declaration
-    type: 'concat :: Vec n
-      (Vec m a) -> Vec (n * m) a'
+    type: 'concat :: Vec n (Vec m a) -> Vec (n * m) a'
     template: |-
       -- concat begin
       ~GENSYM[concat][0] : for ~GENSYM[i][1] in 0 to (~LENGTH[~TYP[0]] - 1) generate
@@ -69,8 +63,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.splitAt
     kind: Expression
-    type: 'splitAt :: SNat
-      m -> Vec (m + n) a -> (Vec m a, Vec n a)'
+    type: 'splitAt :: SNat m -> Vec (m + n) a -> (Vec m a, Vec n a)'
     template: ~IF~LENGTH[~TYPO]~THEN~ARG[1]~ELSE(~VAR[vec][1](0 to ~LIT[0]-1),~VAR[vec][1](~LIT[0]
       to ~VAR[vec][1]'high))~FI
     workInfo: Never
@@ -94,8 +87,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.map
     kind: Declaration
-    type: 'map :: (a -> b)
-      -> Vec n a -> Vec n b'
+    type: 'map :: (a -> b) -> Vec n a -> Vec n b'
     template: |-
       -- map begin
       ~GENSYM[map][0] : for ~GENSYM[i][1] in ~RESULT'range generate~IF ~VIVADO ~THEN~IF~SIZE[~TYP[1]]~THEN
@@ -120,8 +112,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.imap
     kind: Declaration
-    type: 'imap :: KnownNat
-      n => (Index n -> a -> b) -> Vec n a -> Vec n b'
+    type: 'imap :: KnownNat n => (Index n -> a -> b) -> Vec n a -> Vec n b'
     template: |-
       -- imap begin
       ~GENSYM[imap][0] : block
@@ -157,8 +148,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.imap_go
     kind: Declaration
-    type: 'imap_go :: Index
-      n -> (Index n -> a -> b) -> Vec m a -> Vec m b'
+    type: 'imap_go :: Index n -> (Index n -> a -> b) -> Vec m a -> Vec m b'
     template: |-
       -- imap_go begin
       ~GENSYM[imap][5] : for ~GENSYM[i][1] in ~RESULT'range generate~IF ~VIVADO ~THEN~IF~SIZE[~TYP[2]]~THEN
@@ -189,8 +179,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.zipWith
     kind: Declaration
-    type: 'zipWith :: (a ->
-      b -> c) -> Vec n a -> Vec n b -> Vec n c'
+    type: 'zipWith :: (a -> b -> c) -> Vec n a -> Vec n b -> Vec n c'
     template: |-
       -- zipWith begin
       ~GENSYM[zipWith][0] : for ~GENSYM[i][1] in ~RESULT'range generate~IF ~VIVADO ~THEN~IF~SIZE[~TYP[1]]~THEN
@@ -219,8 +208,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.foldr
     kind: Declaration
-    type: 'foldr :: (a ->
-      b -> b) -> b -> Vec n a -> b'
+    type: 'foldr :: (a -> b -> b) -> b -> Vec n a -> b'
     template: |-
       -- foldr begin~IF ~LENGTH[~TYP[2]] ~THEN
       ~GENSYM[foldr][0] : block
@@ -255,8 +243,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.index_int
     kind: Declaration
-    type: 'index_int :: KnownNat
-      n => Vec n a -> Int -> a'
+    type: 'index_int :: KnownNat n => Vec n a -> Int -> a'
     template: |-
       -- index begin
       ~IF~SIZE[~TYP[1]]~THEN~GENSYM[indexVec][0] : block
@@ -275,8 +262,7 @@
     name: Clash.Sized.Vector.replace_int
     kind: Declaration
     outputUsage: NonBlocking
-    type: 'replace_int ::
-      KnownNat n => Vec n a -> Int -> a -> Vec n a'
+    type: 'replace_int :: KnownNat n => Vec n a -> Int -> a -> Vec n a'
     template: |-
       -- replace begin
       ~GENSYM[replaceVec][0] : block
@@ -301,23 +287,20 @@
 - BlackBox:
     name: Clash.Sized.Vector.length
     kind: Expression
-    type: 'length :: KnownNat
-      n => Vec n a -> Int'
+    type: 'length :: KnownNat n => Vec n a -> Int'
     template: to_signed(~LIT[0],~SIZE[~TYPO])
     workInfo: Constant
 - BlackBox:
     name: Clash.Sized.Vector.replicate
     kind: Expression
-    type: 'replicate :: SNat
-      n -> a -> Vec n a'
+    type: 'replicate :: SNat n -> a -> Vec n a'
     template: ~TYPMO'(0 to ~LIT[0]-1 => ~IF ~VIVADO ~THEN ~TOBV[~ARG[1]][~TYP[1]]
       ~ELSE ~ARG[1] ~FI)
     workInfo: Never
 - BlackBox:
     name: Clash.Sized.Vector.transpose
     kind: Declaration
-    type: 'transpose :: KnownNat
-      n => Vec m (Vec n a) -> Vec n (Vec m a)'
+    type: 'transpose :: KnownNat n => Vec m (Vec n a) -> Vec n (Vec m a)'
     template: |-
       -- transpose begin
       ~GENSYM[transpose_outer][2] : for ~GENSYM[row_index][3] in 0 to (~LENGTH[~TYP[1]] - 1) generate
@@ -331,8 +314,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.reverse
     kind: Declaration
-    type: 'reverse :: Vec
-      n a -> Vec n a'
+    type: 'reverse :: Vec n a -> Vec n a'
     template: |-
       -- reverse begin
       ~GENSYM[reverse_loop][2] : for ~GENSYM[i][3] in 0 to (~LENGTH[~TYP[0]] - 1) generate
@@ -371,8 +353,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.rotateLeftS
     kind: Declaration
-    type: 'rotateLeftS ::
-      KnownNat n => Vec n a -> SNat d -> Vec n a'
+    type: 'rotateLeftS :: KnownNat n => Vec n a -> SNat d -> Vec n a'
     template: |-
       -- rotateLeftS begin
       ~GENSYM[rotateLeftS][0] : block
@@ -392,8 +373,7 @@
 - BlackBox:
     name: Clash.Sized.Vector.rotateRightS
     kind: Declaration
-    type: 'rotateRightS ::
-      KnownNat n => Vec n a -> SNat d -> Vec n a'
+    type: 'rotateRightS :: KnownNat n => Vec n a -> SNat d -> Vec n a'
     template: |-
       -- rotateRightS begin
       ~GENSYM[rotateLeftS][0] : block

--- a/clash-lib/prims/vhdl/Clash_Sized_Vector.primitives.yaml
+++ b/clash-lib/prims/vhdl/Clash_Sized_Vector.primitives.yaml
@@ -148,30 +148,30 @@
 - BlackBox:
     name: Clash.Sized.Vector.imap_go
     kind: Declaration
-    type: 'imap_go :: Index n -> (Index n -> a -> b) -> Vec m a -> Vec m b'
+    type: 'imap_go :: (Index n -> a -> b) -> Vec m a -> Index n -> Vec m b'
     template: |-
       -- imap_go begin
-      ~GENSYM[imap][5] : for ~GENSYM[i][1] in ~RESULT'range generate~IF ~VIVADO ~THEN~IF~SIZE[~TYP[2]]~THEN
-        signal ~GENSYM[map_in][2]  : ~TYPEL[~TYP[2]];~ELSE ~FI
+      ~GENSYM[imap][5] : for ~GENSYM[i][1] in ~RESULT'range generate~IF ~VIVADO ~THEN~IF~SIZE[~TYP[1]]~THEN
+        signal ~GENSYM[map_in][2]  : ~TYPEL[~TYP[1]];~ELSE ~FI
         signal ~GENSYM[map_out][3] : ~TYPEL[~TYPO];
-        signal ~GENSYM[i2][4]      : ~TYP[0];
-      begin~IF~SIZE[~TYP[2]]~THEN
-        ~SYM[2] <= fromSLV(~VAR[vec][2](~SYM[1]));~ELSE ~FI
-        ~SYM[4] <= ~ARG[0] + to_unsigned(~SYM[1],~SIZE[~TYP[0]]);
-        ~INST 1
+        signal ~GENSYM[i2][4]      : ~TYP[2];
+      begin~IF~SIZE[~TYP[1]]~THEN
+        ~SYM[2] <= fromSLV(~VAR[vec][1](~SYM[1]));~ELSE ~FI
+        ~SYM[4] <= ~ARG[2] + to_unsigned(~SYM[1],~SIZE[~TYP[2]]);
+        ~INST 0
           ~OUTPUT <= ~SYM[3]~ ~TYPEL[~TYPO]~
-          ~INPUT  <= ~SYM[4]~ ~TYP[0]~
-          ~INPUT  <= ~SYM[2]~ ~TYPEL[~TYP[2]]~
+          ~INPUT  <= ~SYM[4]~ ~TYP[2]~
+          ~INPUT  <= ~SYM[2]~ ~TYPEL[~TYP[1]]~
         ~INST
         ~RESULT(~SYM[1]) <= ~TOBV[~SYM[3]][~TYPEL[~TYPO]];
       end generate;~ELSE
-        signal ~SYM[4] : ~TYP[0];
+        signal ~SYM[4] : ~TYP[2];
       begin
-        ~SYM[4] <= ~ARG[0] + to_unsigned(~SYM[1],~SIZE[~TYP[0]]);
-        ~INST 1
+        ~SYM[4] <= ~ARG[2] + to_unsigned(~SYM[1],~SIZE[~TYP[2]]);
+        ~INST 0
           ~OUTPUT <= ~RESULT(~SYM[1])~ ~TYPEL[~TYPO]~
-          ~INPUT  <= ~SYM[4]~ ~TYP[0]~
-          ~INPUT  <= ~VAR[vec][2](~SYM[1])~ ~TYPEL[~TYP[2]]~
+          ~INPUT  <= ~SYM[4]~ ~TYP[2]~
+          ~INPUT  <= ~VAR[vec][1](~SYM[1])~ ~TYPEL[~TYP[1]]~
         ~INST
       end generate;~FI
       -- imap_go end

--- a/tests/Main.hs
+++ b/tests/Main.hs
@@ -666,6 +666,7 @@ runClashTest = defaultMain
         , runTest "T2831" def{hdlLoad=[],hdlSim=[],hdlTargets=[VHDL]}
         , runTest "T2839" def{hdlLoad=[],hdlSim=[],hdlTargets=[VHDL]}
         , runTest "T2845" def{hdlSim=[],hdlTargets=[Verilog]}
+        , runTest "T2904" def
         ] <>
         if compiledWith == Cabal then
           -- This tests fails without environment files present, which are only

--- a/tests/shouldwork/Issues/T2904.hs
+++ b/tests/shouldwork/Issues/T2904.hs
@@ -1,0 +1,25 @@
+{-# LANGUAGE CPP #-}
+
+module T2904 where
+
+import Clash.Prelude
+import Clash.Explicit.Testbench
+
+topEntity :: Vec 3 (Unsigned 8)
+topEntity = case imap f (5 :> 6 :> 7 :> Nil) of
+  (Cons x xs) -> Cons maxBound xs
+ where
+  f :: Index 3 -> Unsigned 8 -> Unsigned 8
+  f i x = case i of
+    0 -> x + 1
+    1 -> x + 2
+    _ -> x + 3
+{-# CLASH_OPAQUE topEntity #-}
+
+testBench :: Signal System Bool
+testBench = done
+ where
+  done = outputVerifier' clk rst ((maxBound :> 8 :> 10 :> Nil) :> Nil) (pure topEntity)
+  clk  = tbSystemClockGen (not <$> done)
+  rst  = systemResetGen
+{-# CLASH_OPAQUE testBench #-}

